File: 39-CombinationSum.cpp
/*

Given a set of candidate numbers (C) (without duplicates) and a target number (T), find all unique combinations in C where the candidate numbers sums to T.

The same repeated number may be chosen from C unlimited number of times.

Note:
All numbers (including target) will be positive integers.
The solution set must not contain duplicate combinations.
For example, given candidate set [2, 3, 6, 7] and target 7, 
A solution set is: 
[
  [7],
  [2, 2, 3]
]

*/


class Solution {
public:
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        vector<vector<int>> res;
        vector<int> cur;
        // NOTE: gives error without sorting
        sort(candidates.begin(), candidates.end()); 
        dfs(0, candidates, target, cur, res);
        return res;
    }
    
    void dfs(int i, vector<int>& candidates, int target, vector<int>& cur, vector<vector<int>>& res) {
        if(target == 0) {
            res.push_back(cur);
            return;
        }
        for (; i < candidates.size() && target - candidates[i] >= 0; ++i ) {
            cur.push_back(candidates[i]);
            dfs(i, candidates, target - candidates[i], cur, res);
            cur.pop_back();
        }
    }
};

File: java/39-CombinationSum.java
/*

Given a set of candidate numbers (C) (without duplicates) and a target number (T), find all unique combinations in C where the candidate numbers sums to T.

The same repeated number may be chosen from C unlimited number of times.

Note:
All numbers (including target) will be positive integers.
The solution set must not contain duplicate combinations.
For example, given candidate set [2, 3, 6, 7] and target 7, 
A solution set is: 
[
  [7],
  [2, 2, 3]
]

*/


class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> curr = new ArrayList<Integer>();
        if(candidates == null || candidates.length == 0 || target < 0) return res;
        Arrays.sort(candidates);
        helper(res, curr, 0, candidates, target);
        return res;
    }
    private void helper(List<List<Integer>> res, List<Integer> curr, int st, int[] candidates, int target) {
        if(target == 0) res.add(new ArrayList<Integer>(curr));
        else if (target < 0) return;
        else {
            for(int i = st; i < candidates.length; i++) {
                curr.add(candidates[i]);
                helper(res, curr, i, candidates, target-candidates[i]);
                curr.remove((Integer)candidates[i]);
            }
        }
    }
}
